0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+> * 24 rounds of permutation
+
+This section focuses on the actual changes for the proposed gadget.
+
+### Input-Output format
+
+The Keccak gadget is triggered from the sys call. Inputs need to be written into memory and outputs are read from the sys call. One should expect inputs and outputs to be found in bytes serialization, with a big-endian encoding. One can only read from Canon 4 bytes at a time, what means that the hash function will have to wait until the input is fully loaded onto memory.
+
+### Bitwise-sparse representation
+
+Here, we present the bitwise-sparse representation of bitstrings that will be used to obtain more efficient computations of boolean functions within our proof system. 
+
+Given that simple binary operations are very costly within a SNARK, the new idea consists on arithmetizing boolean operations as much as possible, becoming more efficient with finite field algebra. We introduce the _bitwise-sparse representation_ of bitstrings for this goal.
+
+#### Expansion
+
+With this notation, input bitstrings are expanded inserting three `0` bits in between "real" bits (hence the _sparsity_). The motivation for this, is to use intermediate bits to store auxiliary information such as carry bits, resulting from algebraic computations of boolean functions. That means, for each bit in the original bitstring, its expansion will contain four bits. The following expression represents such mapping:
+
+$$\forall X \in \mathbb B^\ell: expand(X) \to 0 \ 0 \ 0 \ b_{\ell-1} \ . \ . \ . \ 0 \ 0 \ 0 \ b_0$$
+
+The reason behind the choice of three empty intermediate bits in the sparse representation follows from the concrete use case of Keccak where no more than $2^4-1$ consecutive boolean operations are performed, and thus carries cannot overwrite the content of the bits to the leftmost bits.
+
+> NOTE: the above means that after each step of the permutation, the expanded representation needs to be reset before starting the next step to ensure completeness of the encoding. 
+
+Even though in each field element of ~254 bits long, one could fit up to 63 real bits with this encoding, each witness value will only store 16 such real bits. Instead, the expansion of the full 64 bit words could be represented by composition of each of the four 16-bit parts. But this will not be computed in the circuit, since the expansion would not fit entirely in the field. The way to perform this mapping is through a 16-bit 2-column lookup table that we call `Reset` (of $2^{16}$ entries, anything significantly larger than that such as 32 bits would be too costly).
+
+Let $sparse(X)$ refer to a representation of $X$, where only the indices in the $4i$-th positions correspond to real bits, and the intermediate indices can contain auxiliary information. Connecting to the above, on input a word $X$, we can expand it in order to perform a series of boolean operations (up to 15) which results in some $sparse(X)$, encoding the same $X$. Then,
```

Comment:
Nit: the notation `sparse(X)` lets the reader think it is a function $sparse : \mathbb{B}^{64} \rightarrow \mathbb{B}^{256}$, where, in reality, it is an element of $\mathbb{B}^{256}$.

Reply:
well, sparse of $X$ means $X$ is something in $\mathbb B^{64}$ and then outputs some element of $\mathbb B^{256}$ that encodes $X$ precisely. I mean $sparse$ is a function, but $sparse(X)$ is referring to the image of the function applied on $X$ so yeah, an element of $\mathbb B^{256}$

Reply:
I am highly in favor of having, for all elements, the set it belongs to. For functions, I wish we can write the domain and image. It is pretty confusing as we talk once about 16 bits, after that 64, after that 256, after that the field F, etc.